src/charm.py

```diff
@@ -95,6 +106,54 @@ def _validate(self):
             if self.config["auth-provider"] == "google":
                 self._check_required_config(REQUIRED_OIDC_CONFIG)
 
+    def _process_wheel_file(self, event):
+        """Process wheel file attached by user.
+
+        This method extracts the wheel file provided by the user and places the contents
+        into the workload container, which can then be read by the Temporal worker.
+
+        Args:
+            event: The event triggered when the relation changed.
+
+        Raises:
+            ValueError: if file is not found.
+        """
+        if not self._state.is_ready():
+            event.defer()
+            return
+
+        try:
+            resource_path = self.model.resources.fetch("workflows-file")
+
+            container = self.unit.get_container(self.name)
+            if not container.can_connect():
+                event.defer()
+                self.unit.status = WaitingStatus("waiting for pebble api")
+                return
+
+            container.exec(["rm", "-rf", "/user_provided"]).wait_output()
+
+            with open(resource_path, "rb") as file:
+                wheel_data = file.read()
+
+                # Push wheel file to the container and extract it.
+                container.push("/user_provided/wheel_file.whl", wheel_data, make_dirs=True)
+                container.exec(["apt-get", "update"]).wait_output()
+                container.exec(["apt-get", "install", "unzip"]).wait_output()
+                container.exec(["unzip", "/user_provided/wheel_file.whl", "-d", "/user_provided"]).wait_output()
+
+                # Find the name of the module provided by the user and set it in state.
+                command = "find /user_provided -mindepth 1 -maxdepth 1 -type d ! -name *.dist-info ! -name *.whl"
+                out, err = container.exec(command.split(" ")).wait_output()
```

Comment:
We should probably handle the case where these execs fail for whatever reason.

Reply:
Updated, thanks!